sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```diff
@@ -0,0 +1,113 @@
+
+
+/**
+ * Generates a [[Projection]] that returns an [[UnsafeRow]].
+ *
+ * Note: The returned UnsafeRow will be pointed to a scratch buffer inside the projection.
+ */
+object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafeProjection] {
+
+  protected def canonicalize(in: Seq[Expression]): Seq[Expression] =
+    in.map(ExpressionCanonicalizer.execute)
+
+  protected def bind(in: Seq[Expression], inputSchema: Seq[Attribute]): Seq[Expression] =
+    in.map(BindReferences.bindReference(_, inputSchema))
+
+  protected def create(expressions: Seq[Expression]): UnsafeProjection = {
+    val ctx = newCodeGenContext()
+    val codes = expressions.map(_.gen(ctx))
+    val all_exprs = codes.map(_.code).mkString("\n")
+
+    val additionalSize = expressions.zipWithIndex.map { case (e, i) =>
+      e.dataType match {
+        case StringType =>
+          s" + (${codes(i).isNull} ? 0 : stringWriter.getSize(${codes(i).primitive}))"
+        case BinaryType =>
+          s" + (${codes(i).isNull} ? 0 : binaryWriter.getSize(${codes(i).primitive}))"
+        case _ => ""
+      }
+    }.mkString("")
+
+    val writers = expressions.zipWithIndex.map { case (e, i) =>
+      val update = e.dataType match {
+        case dt if ctx.isPrimitiveType(dt) =>
+          s"${ctx.setColumn("target", dt, i, codes(i).primitive)}"
+        case StringType =>
+          s"cursor += stringWriter.write(target, ${codes(i).primitive}, $i, cursor)"
+        case BinaryType =>
+          s"cursor += binaryWriter.write(target, ${codes(i).primitive}, $i, cursor)"
+      }
+      s"""if (${codes(i).isNull}) {
+            target.setNullAt($i);
+          } else {
+            $update;
+          }"""
+    }.mkString("\n          ")
+    val fixedSize = (8 * expressions.length) + UnsafeRow.calculateBitSetWidthInBytes(expressions.length)
+
+    val code = s"""
+    public Object generate($exprType[] expr) {
+      return new SpecificProjection();
+    }
+
+    class SpecificProjection extends ${classOf[UnsafeProjection].getName} {
+
+      private final org.apache.spark.sql.catalyst.expressions.StringUnsafeColumnWriter
+        stringWriter = new org.apache.spark.sql.catalyst.expressions.StringUnsafeColumnWriter();
+      private final org.apache.spark.sql.catalyst.expressions.BinaryUnsafeColumnWriter
+        binaryWriter = new org.apache.spark.sql.catalyst.expressions.BinaryUnsafeColumnWriter();
+      private final org.apache.spark.sql.catalyst.expressions.ObjectUnsafeColumnWriter
+        objectWriter = new org.apache.spark.sql.catalyst.expressions.ObjectUnsafeColumnWriter();
+
+      private UnsafeRow target = new UnsafeRow();
+      private byte[] buffer = new byte[64];
+
+      public SpecificProjection() {}
+
+      public Object apply(Object row) {
+        return apply((InternalRow) i);
+      }
+
+      public UnsafeRow apply(InternalRow i) {
+        ${all_exprs}
+
+        int numBytes = $fixedSize $additionalSize;
```

Comment:
Ah, I see that Reynold also got confused by this (https://github.com/apache/spark/pull/7437/files#r34856369), so we should probably take his suggestion to help clarify things here.